content/post/flinksql-time.adoc

```diff
@@ -0,0 +1,1033 @@
+
+[source,javascript]
+----
+Key (14 bytes): {"order_id":1}
+Value (205 bytes):
+    {
+        "order_id": 1,
+        "customer_id": 1001,
+        "total_amount": 149.99,
+        "status": "pending",
+        "created_at": "2025-04-25 09:44:25",
+        "shipped_at": null,
+        "shipping_address": "221B Baker Street, London",
+        "payment_method": "Credit Card"
+    }
+Timestamp: 1745587589625
+----
+
+There are two times here:
+
+. The _event time_ which we'll take as `created_at` from the value: `2025-04-25 09:44:25`.
+. The _processing time_ which is stored in the Kafka message timestamp `1745587589625`.
+This is the milliseconds since the epoch, and https://www.epochconverter.com/[converts] to `2025-04-25 13:26:29.625`.
+
+The timestamp of a Kafka message can be set explicitly https://kafka.apache.org/40/javadoc/org/apache/kafka/clients/producer/ProducerRecord.html[by the producer].
+If it's not set explicitly then the producer sets the timestamp to the current time when it sends the message to the broker.
+The timestamp that's included in the message written by the broker depends on the https://kafka.apache.org/documentation/#topicconfigs_message.timestamp.type[`message.timestamp.type` topic configuration].
+If this is set to `CreateTime` then it uses the time from the producer (either explicit or implicitly set), or it can use the time on the broker when the record is written (`LogAppendTime`).
+
+In theory you could use the Timestamp field of the Kafka message to hold the event time; the data is being stored anyway, so why not optimise by not holding it twice?
+The disadvantage of this comes if you're using a system which doesn't expose the timestamp metadata of a message, so if it doubt, keep it simple :)
+
+== Time in Flink
+
+The longer you've been working with Flink, the higher the chances are that you've heard the word "Watermark".
+I must profess to having spent the last 18 months—since I started working with Flink—with my head in the sand, somewhat Ostrich-like, when it comes to Watermarks.
+They're spoken of in hushed tones and with great reverence.
+They seem to cause great wailing and gnashing of teeth.
+Conference talks are written about them.
+
+I am going to write about Watermarks in this article…just not _quite_ yet.
+
+Before we get to Watermarks, we need to understand that Flink has the concept of a *Time Attribute*.
+There are two types of time attribute:
+
+* Event Time
+* Processing Time (a.k.a. Wall Clock)
+
+Both of these map to the explanations above; when something _happened_ (event time) and when it was _processed_ (…erm, processing time!).
+
+TIP: If you don't care about event time, you can also forget about watermarks.
+Watermarks are an event-time only thing.
+
+In a very rough nut-shell, watermarks define where on the sliding scale between data completeness and data freshness you want your Flink processing to be.
+But like I said, we'll get to watermarks later.
```

Comment:
As a reader, this is somewhat annoying. Why tease me here when not actually talking about it?

Reply:
because I want to walk before we run :) For me, at least, the big bad bogey monster was watermarks, and it dominated my awareness such that I didn't properly understand event/processing time attributes first.